### PR TITLE
fix `np.median` error with MPIArray

### PR DESCRIPTION
**Pull request by ljgray**
…ssed through to the underlying numpy.copy call

### CODE REVIEW

**Comment by ljgray** [comment]
This should fix #205

File: caput/mpiarray.py
Comment (jrs65) [inline caput/mpiarray.py:1389]: As you can guarantee this is properly formed you can probably replace the wrap with `_view_from_data_and_params` which would save some collective communication.


Finally, it *might* be possible to remove this copy method entirely and just reply on using numpys directly. Now we've correctly implemented all the area creation hooks numpys own copy method might just work as-is.

**Review by jrs65 — CHANGES_REQUESTED** [review]

File: caput/mpiarray.py
Comment (ljgray) [reply caput/mpiarray.py:1389]: I'll give both a try

**Comment by ljgray** [comment]
So unfortunately neither of those options fix the issue. The underlying problem seems to be in `__array_finalize__`. Calling `np.median` on an MPIArray calls `__array_finalize__` 3 times. On the second call, the dimension of local_shape is reduced by 1, and on the final call the dimension is reduced to 1. So, if the distributed axis is anything other than 0, the final call will fail because it tries to index into `global_shape`, which has dimension 1, on a higher axis.

**Comment by jrs65** [comment]
I'm confused. I don't why the copy could work using the wrap, but not via `_view_from_data_and_params`. How are you using it? I would have guessed something like this would do it:
```python
return MPIArray. _view_from_data_and_params(
    self.local_array.copy(*args, **kwargs), self.axis, self.global_shape[self.axis], self.local_offset[self.axis], self.comm
)
```
That's essentially what a wrap call does internally, but it skips the checks as you know you don't need them.

**Comment by ljgray** [comment]
> I'm confused. I don't why the copy could work using the wrap, but not via `_view_from_data_and_params`. How are you using it? I would have guessed something like this would do it:
> 
> ```python
> return MPIArray. _view_from_data_and_params(
>     self.local_array.copy(*args, **kwargs), self.axis, self.global_shape[self.axis], self.local_offset[self.axis], self.comm
> )
> ```
> 
> That's essentially what a wrap call does internally, but it skips the checks as you know you don't need them.

Sorry, it doesn't seem to work either way, whether using wrap, _view..., or the numpy copy method. But I don't think that this is actually the issue, as mentioned in my comment above.

**Comment by ljgray** [comment]
I believe that the underlying issue here is a `.ravel()` call in `numpy._median_nancheck`, which doesn't really work for a distributed array.

**Comment by ljgray** [comment]
I think that this should work properly now. The .ravel() call throws a NotImplementedError.

File: caput/mpiarray.py
Comment (jrs65) [inline caput/mpiarray.py:1478]: Can you add a docstring here? Just so that if someone only reads the documentation they'll be able to tell that `.ravel` doesn't work, rather than seeing it's presence and inferring that it doesn't work.

**Review by jrs65 — CHANGES_REQUESTED** [review]
Looks good to me, I think just the docstring comment.

Also, can you rebase all the commits down to just the externally relevant changes before merging?

File: caput/mpiarray.py
Comment (ljgray) [reply caput/mpiarray.py:1478]: Oh, good catch.

File: caput/mpiarray.py
Comment (jrs65) [inline caput/mpiarray.py:1468]: Also there's a formatting issue in this docstring.

**Review by jrs65 — CHANGES_REQUESTED** [review]

**Pull request merged**
